File: 2416-evaluate-boolean-binary-tree/evaluate-boolean-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
       
        def rec(root):


            if (root.left  == None) and (root.right == None):
                return bool(root.val)
            l ,r  = rec(root.left),rec(root.right)
         
            if root.val ==2 :
                ans =  (l or  r)
            
            if root.val == 3:
                ans  = (l and  r)
            
            return ans
        
        return rec(root)

        

File: 2416-evaluate-boolean-binary-tree/evaluate-boolean-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean evaluateTree(TreeNode root) {
        return rec(root);
    }
     private boolean rec(TreeNode root) {
        if (root.left == null && root.right == null) {
            return root.val != 0;
        }
        boolean l = rec(root.left);
        boolean r = rec(root.right);
        boolean ans = false;
        if (root.val == 2) {
            ans = l || r;
        }
        if (root.val == 3) {
            ans = l && r;
        }
        return ans;
    }
}